kmod/src/quorum.c

```diff
@@ -13,766 +13,1073 @@
-	struct scoutfs_inet_addr addr;
-	enum { VOTER, CANDIDATE };
-	ktime_t cycle_to;
-	ktime_t term_to;
-	LIST_HEAD(blocks);
-	u64 vote_for_write_nr;
-	u64 vote_for_rid;
-	u64 write_nr;
-	u64 term;
-	int log_cycles = 0;
-	int votes;
-	int role;
+	struct scoutfs_quorum_block blk;
+	struct sockaddr_in unused;
+	struct quorum_host_msg msg;
+	struct quorum_status qst;
+	__le64 mark;
+	u64 blkno;
 	int ret;
 
-	*elected_term = 0;
-
-	trace_scoutfs_quorum_election(sb, prev_term);
+	/* recording votes from slots as native single word bitmap */
+	BUILD_BUG_ON(SCOUTFS_QUORUM_MAX_SLOTS > BITS_PER_LONG);
 
-	super = kmalloc(sizeof(struct scoutfs_super_block), GFP_NOFS);
-	our_blk = kmalloc(SCOUTFS_BLOCK_SM_SIZE, GFP_NOFS);
-	if (!super || !our_blk) {
-		ret = -ENOMEM;
+	/* get our starting term from our persistent block */
+	mark = 0;
+	blkno = SCOUTFS_QUORUM_BLKNO + opts->quorum_slot_nr;
+	ret = read_quorum_block(sb, blkno, &blk, &mark);
+	if (ret < 0)
 		goto out;
-	}
-
-	/* start out as a passive voter */
-	role = VOTER;
-	term = 0;
-	write_nr = 0;
-	vote_for_rid = 0;
-	vote_for_write_nr = 0;
-
-	/* we'll become a candidate if we don't see another candidate */
-	term_to = random_to(SCOUTFS_QUORUM_TERM_LO_MS,
-			    SCOUTFS_QUORUM_TERM_HI_MS);
 
-	for (;;) {
-		memset(our_blk, 0, SCOUTFS_BLOCK_SM_SIZE);
+	/* start out as a follower */
+	qst.role = FOLLOWER;
+	qst.term = le64_to_cpu(blk.term);
+	qst.vote_for = -1;
+	qst.vote_bits = 0;
 
-		scoutfs_inc_counter(sb, quorum_cycle);
+	/* see if there's a server to chose heartbeat or election timeout */
+	if (scoutfs_quorum_server_sin(sb,&unused) == 0)
+		qst.timeout = heartbeat_timeout();
+	else
+		qst.timeout = election_timeout();
 
-		ret = scoutfs_read_super(sb, super);
-		if (ret)
-			goto out;
+	while (!qinf->shutdown) {
 
-		/* done if we see evidence of a new server */
-		if (le64_to_cpu(super->quorum_server_term) > prev_term) {
-			scoutfs_inc_counter(sb, quorum_saw_super_leader);
+		ret = recv_msg(sb, &msg, qst.timeout);
+		if (ret < 0) {
+			if (ret != -ETIMEDOUT && ret != -EAGAIN) {
+				scoutfs_err(sb, "quorum recvmsg err %d", ret);
+				scoutfs_inc_counter(sb, quorum_recv_error);
+				goto out;
+			}
+			msg.type = SCOUTFS_QUORUM_MSG_INVALID;
 			ret = 0;
-			goto out;
 		}
 
-		/* done if we couldn't elect anyone */
-		if (ktime_after(ktime_get(), timeout_abs)) {
-			scoutfs_inc_counter(sb, quorum_timedout);
-			ret = -ETIMEDOUT;
-			goto out;
+		/* ignore messages from older terms */
+		if (msg.type != SCOUTFS_QUORUM_MSG_INVALID &&
+		    msg.term < qst.term)
+			msg.type = SCOUTFS_QUORUM_MSG_INVALID;
+
+		/* if the server has shutdown we become follower */
+		if (!test_bit(QINF_FLAG_SERVER, &qinf->flags) &&
+		    qst.role == LEADER) {
+			qst.role = FOLLOWER;
+			qst.vote_for = -1;
+			qst.vote_bits = 0;
+			qst.timeout = election_timeout();
+			scoutfs_inc_counter(sb, quorum_server_shutdown);
+
+			send_msg_others(sb, SCOUTFS_QUORUM_MSG_RESIGNATION,
+					qst.term);
```

Comment:
Do we need resignation?

Reply:
I think so.  Without it quorum members can't differentiate between a server dying, so they wait for the longer heartbeat timeout, and a server exiting cleanly, at which point they should use the shorter election timeout.  Without it umount of the leader would cause a heartbeat timeout delay in service until the next leader is elected.  The mount/unmount test would barely do anything, it would be spending all its time sitting around waiting.  This is one of the effects of our different model than the share-nothing raft key/value store.. we're using the leader election to stand up and shut down an exclusive service.